File: src/test/java/uk/gov/justice/hmpps/prison/security/AuthenticationFacadeTest.java
package uk.gov.justice.hmpps.prison.security;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.CsvSource;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.oauth2.jwt.Jwt;
import uk.gov.justice.hmpps.prison.web.config.AuthAwareAuthenticationToken;

import java.util.Set;

import static java.util.Collections.emptySet;
import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.mock;
import static uk.gov.justice.hmpps.prison.security.AuthSource.AUTH;
import static uk.gov.justice.hmpps.prison.security.AuthSource.NOMIS;
import static uk.gov.justice.hmpps.prison.security.AuthSource.NONE;

public class AuthenticationFacadeTest {
    private final AuthenticationFacade authenticationFacade = new AuthenticationFacade();

    @Test
    public void getAuthenticationSource_AuthSource_nomis() {
        setAuthentication("nomis");
        assertThat(authenticationFacade.getAuthenticationSource()).isEqualTo(NOMIS);
    }

    @Test
    public void getProxyUserAuthenticationSource_AuthSource_auth() {
        setAuthentication("auth");
        assertThat(authenticationFacade.getAuthenticationSource()).isEqualTo(AUTH);
    }

    @Test
    public void getProxyUserAuthenticationSource_AuthSource_null() {
        setAuthentication(null);
        assertThat(authenticationFacade.getAuthenticationSource()).isEqualTo(NONE);
    }

    @Test
    public void getProxyUserAuthenticationSource_NoUserAuthentication() {
        SecurityContextHolder.getContext().setAuthentication(null);
        assertThat(authenticationFacade.getAuthenticationSource()).isEqualTo(NONE);
    }

    @ParameterizedTest
    @CsvSource({
        "ROLE_SYSTEM_USER,true",
        "SYSTEM_USER,true",
        "SYSTEMUSER,false"
    })
    public void hasRolesTest(String role, boolean expected) {
        setAuthentication("auth", Set.of(
            new SimpleGrantedAuthority("ROLE_SYSTEM_USER")
        ));

        assertThat(authenticationFacade.hasRoles(role)).isEqualTo(expected);
        assertThat(authenticationFacade.isOverrideRole(role)).isEqualTo(expected);
    }

    private void setAuthentication(final String source) {
        setAuthentication(source, emptySet());
    }

    private void setAuthentication(final String source, Set<GrantedAuthority> authoritySet) {
        final Authentication auth = new AuthAwareAuthenticationToken(mock(Jwt.class), "client", source, authoritySet);
        SecurityContextHolder.getContext().setAuthentication(auth);
    }
}

File: src/test/java/uk/gov/justice/hmpps/prison/security/AuthenticationFacadeTest.kt
package uk.gov.justice.hmpps.prison.security

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.CsvSource
import org.mockito.Mockito.mock
import org.springframework.security.core.Authentication
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.security.oauth2.jwt.Jwt
import uk.gov.justice.hmpps.kotlin.auth.AuthAwareAuthenticationToken
import uk.gov.justice.hmpps.kotlin.auth.AuthSource
import uk.gov.justice.hmpps.kotlin.auth.AuthSource.AUTH
import uk.gov.justice.hmpps.kotlin.auth.AuthSource.NOMIS
import uk.gov.justice.hmpps.kotlin.auth.AuthSource.NONE

class AuthenticationFacadeTest {
  private val authenticationFacade = AuthenticationFacade()

  @Test
  fun getAuthenticationSource_AuthSource_nomis() {
    setAuthentication(NOMIS)
    assertThat(authenticationFacade.authenticationSource).isEqualTo(NOMIS)
  }

  @Test
  fun getProxyUserAuthenticationSource_AuthSource_auth() {
    setAuthentication(AUTH)
    assertThat(authenticationFacade.authenticationSource).isEqualTo(AUTH)
  }

  @Test
  fun proxyUserAuthenticationSource_NoUserAuthentication() {
    SecurityContextHolder.getContext().authentication = null
    assertThat(authenticationFacade.authenticationSource).isEqualTo(NONE)
  }

  @ParameterizedTest
  @CsvSource("ROLE_SYSTEM_USER,true", "SYSTEM_USER,true", "SYSTEMUSER,false")
  fun hasRolesTest(role: String, expected: Boolean) {
    setAuthentication(
      AUTH,
      setOf<GrantedAuthority>(
        SimpleGrantedAuthority("ROLE_SYSTEM_USER"),
      ),
    )
    assertThat(AuthenticationFacade.hasRoles(role)).isEqualTo(expected)
    assertThat(authenticationFacade.isOverrideRole(role)).isEqualTo(expected)
    assertThat(authenticationFacade.isClientOnly).isFalse
  }

  @ParameterizedTest
  @CsvSource("ROLE_SYSTEM_USER,true", "SYSTEM_USER,true", "SYSTEMUSER,false")
  fun hasClientRolesTest(role: String, expected: Boolean) {
    setAuthentication(
      AUTH,
      setOf<GrantedAuthority>(
        SimpleGrantedAuthority("ROLE_SYSTEM_USER"),
      ),
      userName = null,
    )
    assertThat(AuthenticationFacade.hasRoles(role)).isEqualTo(expected)
    assertThat(authenticationFacade.isOverrideRole(role)).isEqualTo(expected)
    assertThat(authenticationFacade.isClientOnly).isTrue
  }

  private fun setAuthentication(source: AuthSource) {
    setAuthentication(source, emptySet())
  }

  private fun setAuthentication(source: AuthSource, authoritySet: Set<GrantedAuthority>, userName: String? = "userName") {
    val auth: Authentication = AuthAwareAuthenticationToken(mock(Jwt::class.java), "clientId", userName, source, authoritySet)
    SecurityContextHolder.getContext().authentication = auth
  }

  @Test
  fun isOverrideRole_NoOverrideRoleSet() {
    assertThat(authenticationFacade.isOverrideRole()).isFalse()
  }

  @Test
  fun hasRoles_NoAllowedRoleSet() {
    assertThat(AuthenticationFacade.hasRoles()).isFalse()
  }

  @Test
  fun getClientId() {
    setAuthentication(NONE)
    assertThat(authenticationFacade.clientId).isEqualTo("clientId")
  }
}
